value/vector.go

```diff
@@ -379,82 +397,84 @@ func (v Vector) grade(c Context) Vector {
-func (v Vector) inverse(c Context) Value {
-	if len(v) == 0 {
+func (v *Vector) inverse(c Context) Value {
+	if v.Len() == 0 {
 		Errorf("inverse of empty vector")
 	}
-	if len(v) == 1 {
+	if v.Len() == 1 {
 		return inverse(c, v)
 	}
 	// We could do this evaluation using "conj" and "+.*" but avoid the overhead.
-	conj := v.Copy().(Vector)
-	for i, x := range conj {
+	conj := v.Copy().(*Vector)
+	for i, x := range conj.All() {
 		if !IsScalarType(x) {
 			Errorf("inverse of vector with non-scalar element")
 		}
 		if cmplx, ok := x.(Complex); ok {
-			conj[i] = NewComplex(cmplx.real, c.EvalUnary("-", cmplx.imag)).shrink()
+			conj.Set(i, NewComplex(cmplx.real, c.EvalUnary("-", cmplx.imag)).shrink())
 		}
 	}
 	mag := Value(zero)
-	for i, x := range v {
-		mag = c.EvalBinary(mag, "+", c.EvalBinary(x, "*", conj[i]))
+	for i, x := range v.All() {
+		mag = c.EvalBinary(mag, "+", c.EvalBinary(x, "*", conj.At(i)))
 	}
 	if isZero(mag) {
 		Errorf("inverse of zero vector")
 	}
-	for i, x := range conj {
-		conj[i] = c.EvalBinary(x, "/", mag)
+	for i, x := range conj.All() {
+		conj.Set(i, c.EvalBinary(x, "/", mag))
 	}
 	return conj
 }
 
 // membership creates a vector of size len(u) reporting
 // whether each element of u is an element of v.
 // Algorithm is O(nV log nV + nU log nV) where nU==len(u) and nV==len(V).
-func membership(c Context, u, v Vector) []Value {
-	values := make([]Value, len(u))
+func membership(c Context, u, v *Vector) []Value {
+	values := make([]Value, u.Len())
 	sortedV := v.sortedCopy(c)
-	work := 2 * (1 + int(math.Log2(float64(len(v)))))
+	work := 2 * (1 + int(math.Log2(float64(v.Len()))))
 	pfor(true, work, len(values), func(lo, hi int) {
 		for i := lo; i < hi; i++ {
-			values[i] = toInt(sortedV.contains(c, u[i]))
+			values[i] = toInt(sortedV.contains(c, u.At(i)))
 		}
 	})
 	return values
 }
 
 // sortedCopy returns a copy of v, in ascending sorted order.
-func (v Vector) sortedCopy(c Context) Vector {
-	sortedV := make([]Value, len(v))
-	copy(sortedV, v)
+func (v *Vector) sortedCopy(c Context) *Vector {
+	sortedV := make([]Value, v.Len())
+	copy(sortedV, v.All())
 	sort.Slice(sortedV, func(i, j int) bool {
 		return OrderedCompare(c, sortedV[i], sortedV[j]) < 0
 	})
-	return sortedV
+	return NewVector(sortedV)
 }
 
 // contains reports whether x is in v, which must be already in ascending
 // sorted order.
-func (v Vector) contains(c Context, x Value) bool {
-	pos := sort.Search(len(v), func(j int) bool {
-		return OrderedCompare(c, v[j], x) >= 0
+func (v *Vector) contains(c Context, x Value) bool {
+	pos := sort.Search(v.Len(), func(j int) bool {
+		return OrderedCompare(c, v.At(j), x) >= 0
 	})
-	return pos < len(v) && OrderedCompare(c, v[pos], x) == 0
+	return pos < v.Len() && OrderedCompare(c, v.At(pos), x) == 0
 }
 
-func (v Vector) shrink() Value {
-	if len(v) == 1 {
-		return v[0]
+func (v *Vector) shrink() Value {
+	if v.Len() == 1 {
+		return v.At(0)
```

Comment:
I would just return v.data[0]. This is internal code. 
Generally throughout this file I'd do that rather than use the method. Did all these indexing changes come due to automation or do you have a philosophical reason to go the hard way?

It's OK as is, and I'm prepared to merge, but I'm curious why you went this route. The reverse method for instance seems much clumsier.

Reply:
The idea is to make it easier to change the internal data representation away from a simple slice at some point in the future. I have an experimental change where a vector is a radix-32 trie, which means that if you have something like

```
m = 100 100 rho iota 10000
x = m
x[20;30] = -1
```

then x and m can share most of their storage. This is particularly helpful if you have operator functions that take a big matrix, update a few elements, and pass it back, since it ends up being O(1) copying instead of O(n) copying.

Reply:
OK.